Cubical/HITs/Cylinder.agda

```diff
@@ -7,65 +7,263 @@ open import Cubical.Core.Everything
+-- constitutes a cylinder functor:
+--
+--   https://ncatlab.org/nlab/show/cylinder+functor
+--
+-- e₀ = inl
+-- e₁ = inr
+-- σ = out
+module Functorial where
+  private
+    variable
+      ℓa ℓb ℓc : Level
+      A : Set ℓa
+      B : Set ℓb
+      C : Set ℓc
+
+  mapCylinder : (A → B) → Cylinder A → Cylinder B
+  mapCylinder f (inl x) = inl (f x)
+  mapCylinder f (inr x) = inr (f x)
+  mapCylinder f (cross x i) = cross (f x) i
+
+  mapCylinderId : mapCylinder (λ(x : A) → x) ≡ (λ x → x)
+  mapCylinderId i (inl x) = inl x
+  mapCylinderId i (inr x) = inr x
+  mapCylinderId i (cross x j) = cross x j
+
+  mapCylinder∘
+    : (f : A → B) → (g : B → C)
+    → mapCylinder (λ x → g (f x)) ≡ (λ x → mapCylinder g (mapCylinder f x))
+  mapCylinder∘ f g i (inl x) = inl (g (f x))
+  mapCylinder∘ f g i (inr x) = inr (g (f x))
+  mapCylinder∘ f g i (cross x j) = cross (g (f x)) j
+
+  -- There is an adjunction between the cylinder and coyclinder
+  -- functors.
+  --
+  --   Cylinder ⊣ Cocylinder
+  adj₁ : (Cylinder A → B) → A → Cocylinder B
+  adj₁ f x zero = f (inl x)
+  adj₁ f x one = f (inr x)
+  adj₁ f x (seg i) = f (cross x i)
+
+  adj₂ : (A → Cocylinder B) → Cylinder A → B
+  adj₂ g (inl x) = g x zero
+  adj₂ g (inr x) = g x one
+  adj₂ g (cross x i) = g x (seg i)
+
+  adj₁₂ : (g : A → Cocylinder B) → adj₁ (adj₂ g) ≡ g
+  adj₁₂ g _ x zero = g x zero
+  adj₁₂ g _ x one = g x one
+  adj₁₂ g _ x (seg i) = g x (seg i)
+
+  adj₂₁ : (f : Cylinder A → B) → adj₂ (adj₁ f) ≡ f
+  adj₂₁ f j (inl x) = f (inl x)
+  adj₂₁ f j (inr x) = f (inr x)
+  adj₂₁ f j (cross x i) = f (cross x i)
+
+module IntervalEquiv where
+  -- There is an equivalence between the interval and the
+  -- cylinder over the unit type.
+  Interval→CylinderUnit : Interval → Cylinder Unit
+  Interval→CylinderUnit zero = inl _
+  Interval→CylinderUnit one = inr _
+  Interval→CylinderUnit (seg i) = cross _ i
+
+  CylinderUnit→Interval : Cylinder Unit → Interval
+  CylinderUnit→Interval (inl _) = zero
+  CylinderUnit→Interval (inr _) = one
+  CylinderUnit→Interval (cross _ i) = seg i
+
+  Interval→CylinderUnit→Interval
+    : ∀ i → CylinderUnit→Interval (Interval→CylinderUnit i) ≡ i
+  Interval→CylinderUnit→Interval zero = refl
+  Interval→CylinderUnit→Interval one = refl
+  Interval→CylinderUnit→Interval (seg i) = refl
+
+  CylinderUnit→Interval→CylinderUnit
+    : ∀ c → Interval→CylinderUnit (CylinderUnit→Interval c) ≡ c
+  CylinderUnit→Interval→CylinderUnit (inl _) = refl
+  CylinderUnit→Interval→CylinderUnit (inr _) = refl
+  CylinderUnit→Interval→CylinderUnit (cross _ i) = refl
+
+  CylinderUnit≃Interval : Cylinder Unit ≃ Interval
+  CylinderUnit≃Interval =
+    isoToEquiv
+      CylinderUnit→Interval
+      Interval→CylinderUnit
+      Interval→CylinderUnit→Interval
+      CylinderUnit→Interval→CylinderUnit
+
+
+  -- More generally, there is an equivalence between the cylinder
+  -- over any type A and the product of A and the interval.
+  module _ {ℓ} {A : Set ℓ} where
```

Comment:
This `ℓ` and `A` can be a `variable` too, I believe.

Reply:
What is the point of this, though? It's just adding 4 extra lines for something that already fits easily on an existing line.

Reply:
Oh, also, it actually can't be made a variable. I forgot, but I tried it, and Agda complained about it being used in e.g. `Cyl` below.

Reply:
@ice1000 I don't think we should push `variable` as much as you are suggesting. Modules are great at abstracting some parameters out of a bunch of definitions. 

In general I would use `variable` to establish specific naming conventions, not for local arguments.

>  I just want to avoid indenting definitions

Do you have a specific reason for that? Other than saving horizontal space?

Reply:
> @ice1000 I don't think we should push variable as much as you are suggesting.

Yes, I agree. My final suggestion is to push the levels to `variable` but Agda does not support using `variable`s in GADT definitions, which has probably confused OP.

> Do you have a specific reason for that?

Yes,

> Other than saving horizontal space?

Then no.